api javax.* => jakarta.* and added top level dependency on jakarta.interceptor

scottmarlow opened the pull request:
Signed-off-by: Scott Marlow <smarlow@redhat.com>

This should be able to replace [jta-api/pull#72](https://github.com/eclipse-ee4j/jta-api/pull/72)

tomjenkinson commented on api/pom.xml:
```diff
@@ -312,7 +312,13 @@
+        <dependency>
+            <groupId>jakarta.interceptor</groupId>
+            <artifactId>jakarta.interceptor-api</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
```
The failure in CI is initially caused by these dependencies not existing in the repo we are reading from. Do you know if these snapshots have been pushed to a repo yet? If not, perhaps we can request the projects [1] [2] if they are able to share snapshots yet on https://jakarta.oss.sonatype.org/content/repositories/snapshots/

[1] https://accounts.eclipse.org/mailing-list/cdi-dev
[2] https://accounts.eclipse.org/mailing-list/interceptors-dev

scottmarlow replied:
There is supposed to be sharing of Jakarta snapshots soon.  From [Migration to Jakarta installation of Nexus](https://www.eclipse.org/lists/ee4j-pmc/msg02361.html) thread, it is still work in progress.  I think that a certain (Eclipse) parent pom needs to be used as well.  I listened in on the Jakarta EE 9 platform call today and this was discussed (briefly) today.

tomjenkinson commented on api/pom.xml:
```diff
@@ -27,12 +27,12 @@
     <groupId>jakarta.transaction</groupId>
     <artifactId>jakarta.transaction-api</artifactId>
-    <version>1.3.4-SNAPSHOT</version>
+    <version>2.0-SNAPSHOT</version>
```
I think this would be as well in a separate commit

tomjenkinson replied:
@scottmarlow I think this is the place that @m0mus identified over here: https://github.com/eclipse-ee4j/jta-api/pull/72#discussion_r366273668 where the version should be 2.0.0-SNAPSHOT

tomjenkinson commented on api/pom.xml:
```diff
@@ -27,12 +27,12 @@
         <non.final>false</non.final>
         <extension.name>javax.transaction</extension.name>
-        <spec.version>1.3</spec.version>
+        <spec.version>2.0</spec.version>
```
I think this would be as well in the commit for the 2.0-SNAPSHOT pom change. Perhaps we should change the root pom.xml in that commit too: https://github.com/eclipse-ee4j/jta-api/blob/master/pom.xml#L31

scottmarlow replied:
For the root pom.xml change, should we use 2.0.0, so something like the following in root pom.xml + spec/pom.xml:
>   <artifactId>jakarta.transaction-parent</artifactId>
>  <version>2.0.0-SNAPSHOT</version>
 

Note that api/pom.xml has a different parent:

>         <groupId>org.eclipse.ee4j</groupId>
>         <artifactId>project</artifactId>
>         <version>1.0.5</version>

tomjenkinson replied:
I think so, as that is the overall version of the project that we aim at.

scottmarlow replied:
Makes sense

tomjenkinson commented on api/pom.xml:
```diff
@@ -157,7 +157,7 @@
                         <Implementation-Vendor>${project.organization.name}</Implementation-Vendor>
                         <Implementation-Vendor-Id>org.glassfish</Implementation-Vendor-Id>
-                        <Import-Package>javax.interceptor.*;version="[1.2,1.2.98)",*</Import-Package>
+                        <Import-Package>jakarta.interceptor.*;version="[1.2,2.0.0)",*</Import-Package>
```
I think this would probably be in the same commit as the 2.0-SNAPSHOT change

tomjenkinson replied:
I think this was already resolved?

tomjenkinson commented on api/pom.xml:
```diff
@@ -312,7 +312,13 @@
+            <version>3.0.0-SNAPSHOT</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
```
I think adding this dependency would be a third commit

tomjenkinson replied:
the reason it is not in commit 4 is because it addresses the enhancement to add the core dependency on interceptor

tomjenkinson commented on api/pom.xml:
```diff
@@ -312,7 +312,13 @@
             <groupId>jakarta.enterprise</groupId>
             <artifactId>jakarta.enterprise.cdi-api</artifactId>
-            <version>2.0.1</version>
+            <version>3.0.0-SNAPSHOT</version>
```
EDIT: I think this would be in the fourth commit related to dependency upgrades for jakarta.*

tomjenkinson commented:
Thanks for the contribution.

I am thinking of 4 commits in this PR:
1. Just the code package change of Jakarta transactions (i.e. not the CDI/interceptor) package changes)
2. The version bump in our poms to 2.0 // I think this is important once commit 1 is merged but can be done in a separate commit
3. The update of CDI code and the pom version bump for CDI, plus the interceptor code changes and Import-Package change (not the addition of the direct dependency in the pom though) related to jakarta.*
4. The addition of a direct dependency on interceptor // I think this is tidy up and useful for clarity

The main ordering constraint would be that #2 should be after #1 and #4 should be after #3 (due to the version number addition).

tomjenkinson commented on api/src/main/java/jakarta/transaction/Transactional.java:
```diff
@@ -14,14 +14,14 @@
-import javax.enterprise.util.Nonbinding;
-import javax.interceptor.InterceptorBinding;
+import jakarta.enterprise.util.Nonbinding;
+import jakarta.interceptor.InterceptorBinding;
```
This would be in the dependency upgrade commit

tomjenkinson commented on api/src/main/java/jakarta/transaction/TransactionScoped.java:
```diff
@@ -53,7 +53,7 @@
  * The contextual references used across different Jakarta Transactions transactions are distinct.
  * Refer to the CDI specification for more details on contextual references.
- * A <code>javax.enterprise.context.ContextNotActiveException</code>
+ * A <code>jakarta.enterprise.context.ContextNotActiveException</code>
```
This would be in the dependency upgrade commit

tomjenkinson commented on api/src/main/java/jakarta/transaction/TransactionScoped.java:
```diff
@@ -14,16 +14,16 @@
 import java.lang.annotation.Target;
 
-import javax.enterprise.context.NormalScope;
+import jakarta.enterprise.context.NormalScope;
```
This would be in the dependency upgrade commit

scottmarlow replied:
I will edit the commit and move it.

scottmarlow replied:
Wait, which of the following commits do you consider the dependency upgrade commit?

> 1. Just the code package change of Jakarta transactions (i.e. not the CDI/interceptor) package changes)
> 2. The version bump in our poms to 2.0 // I think this is important once commit 1 is merged but can be done in a separate commit
> 3. The update of CDI code and the pom version bump for CDI, plus the interceptor code changes and Import-Package change (not the addition of the direct dependency in the pom though) related to jakarta.*
> 4. The addition of a direct dependency on interceptor // I think this is tidy up and useful for clarity5.

tomjenkinson replied:
I think this was already resolved, sorry for the confusion

tomjenkinson commented on api/pom.xml:
```diff
@@ -312,7 +312,13 @@
+        </dependency>
+        <dependency>
+            <groupId>jakarta.interceptor</groupId>
+            <artifactId>jakarta.interceptor-api</artifactId>
```
I think adding this direct dependency would be a separate enhancement commit

scottmarlow replied:
It is a separate commit, do you mean to remove this change from this pull request?

tomjenkinson replied:
I think this was already resolved?

scottmarlow commented:
>     1. Just the code package change of Jakarta transactions (i.e. not the CDI/interceptor) package changes)
I reverted the other changes and squashed into the first commit.  Other changes will follow.

tomjenkinson commented:
Thanks Scott, I wasn't suggesting to not include them in this PR BTW, just to be able to review them as separate commits.

tomjenkinson commented:
retest this please

scottmarlow commented:
@tomjenkinson My first attempt was to "git reset" the one commit and then break it into separate commits, however, since that turned into:
- Deletes of the original javax.* source files, losing all history :(
- New jakarta.* source files with no history from the original javax.* :(

I decided it was easier to revert the changes that shouldn't be in the first commit and will apply them in separate commits, as you requested.  Perhaps there is some other git magic that I could of used (I did web search and found some suggestions but nothing sounded easy, perhaps I missed something).

tomjenkinson commented:
Thanks @scottmarlow, would you like me to review this just as it is or are you going to add the further commits to this PR? The actual package change commit itself looks good. The CI is failing because we move on to testing with glassfish and that can't find the javax.transaction.* imports any more (as could be expected)

scottmarlow commented:
@tomjenkinson please do review the four commits.  Thanks!

tomjenkinson commented:
The diffs look great thanks @scottmarlow

From that point it looks good but I think we should try to work out how if we can get CI a bit further on. Given the causes of the failures, do you know when we might get snapshots of the dependencies?

scottmarlow commented:
> do you know when we might get snapshots of the dependencies?

I'm not sure but I think https://www.eclipse.org/lists/ee4j-pmc/msg02412.html is the email thread that discusses the migration to the Jakarta nexus repo.  I think that there is some discussion about how to deploy snapshots to the new repo.  

Also, there is an [EE 9 board](https://github.com/orgs/eclipse-ee4j/projects/17#column-7346425) for tracking progress.

tomjenkinson commented:
Thanks @scottmarlow - at the moment we are blocked on:

https://github.com/eclipse-ee4j/cdi/issues/416
https://github.com/eclipse-ee4j/interceptor-api/issues/53

antoinesd commented:
@tomjenkinson CDI `javax.*` package were switched to `jakarta.*`. Is there something else blocking for you on CDI side?

tomjenkinson commented:
Thanks @antoinesd - it would enable us to verify [this commit](https://github.com/eclipse-ee4j/jta-api/pull/74/commits/87cb14bf9bd31c9ecd9847617f47489fa46b3cc4) if you were to have a milestone deployed? Is that possible yet?

tomjenkinson commented:
@scottmarlow please can you rebase your branch?

antoinesd commented:
@tomjenkinson we could but are blocked by the lack of milestone or snapshot Interceptor api

tomjenkinson commented:
Right, thanks for the info @antoinesd

scottmarlow commented:
@tomjenkinson just rebased and pushed.

tomjenkinson commented:
Thanks @scottmarlow for rebasing

tomjenkinson commented:
We are still waiting for the milestone builds of CDI and interceptor which will require changes in at least [here](https://github.com/eclipse-ee4j/jta-api/pull/74/commits/754d2c4350b38655a8fc17caf68bf6108dac6212#diff-8658bb96fff6ac848cd28d3242fce067R287) and [here](https://github.com/eclipse-ee4j/jta-api/pull/74/commits/3e1917c04f0344c76a4d8b2cd4d88d4b1150f819#diff-8658bb96fff6ac848cd28d3242fce067R293)

tomjenkinson commented:
Just to be clear @scottmarlow I think we are OK until we get milestones then your commits would just move to using those.

tomjenkinson commented:
@scottmarlow interceptors seem to have released an RC1, please could you update your PR to pull in that version? https://github.com/eclipse-ee4j/interceptor-api/issues/61

tomjenkinson commented:
@antoinesd FYI too https://github.com/eclipse-ee4j/interceptor-api/issues/61

scottmarlow commented:
@tomjenkinson sure, just pushed a change to use interceptor 2.0.0-RC1

tomjenkinson commented:
Thanks @scottmarlow. @starksm64 or @antoinesd , would you be able to send a message on this PR when the CDI milestone is available, please? @scottmarlow when that is available, would you be able to push a new version of your [patch](https://github.com/eclipse-ee4j/jta-api/pull/74/commits/754d2c4350b38655a8fc17caf68bf6108dac6212) to reference that version please?

starksm64 commented:
Sure, I will do a PR once we have the 3.0.0-M1 milestone released.

tomjenkinson commented:
Thanks @starksm64 but I think if you just let us know then @scottmarlow might be best to edit his existing commit to reference this version because updating to the CDI version is part of a set of related changes that will be as well to test all in one place. You can see the changes expected here https://github.com/eclipse-ee4j/jta-api/pull/74/commits/754d2c4350b38655a8fc17caf68bf6108dac6212. 

My hope is if @scottmarlow just changes 3.0.0-SNAPSHOT to 3.0.0-M1 (or whatever the version of the CDI API artifact is that is that is released) our CI will get further. Perhaps if you have confidence of the version number already then @scottmarlow could already edit the commit and we can just trigger CI once the artifact is available in the repo?

starksm64 commented:
Ok, ack, yes the version will be 3.0.0-M1.

On Fri, Feb 14, 2020 at 12:54 PM Tom Jenkinson <notifications@github.com>
wrote:

> Thanks @starksm64 <https://github.com/starksm64> but I think if you just
> let us know then @scottmarlow <https://github.com/scottmarlow> might be
> best to edit his existing commit to reference this version because updating
> to the CDI version is part of a set of related changes that will be as well
> to test all in one place. You can see the changes expected here 754d2c4
> <https://github.com/eclipse-ee4j/jta-api/commit/754d2c4350b38655a8fc17caf68bf6108dac6212>
> .
>
> My hope is if @scottmarlow <https://github.com/scottmarlow> just changes
> 3.0.0-SNAPSHOT to 3.0.0-M1 (or whatever the version of the CDI API artifact
> is that is that is released) our CI will get further. Perhaps if you have
> confidence of the version number already then @scottmarlow
> <https://github.com/scottmarlow> could already edit the commit and we can
> just trigger CI once the artifact is available in the repo?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-ee4j/jta-api/pull/74?email_source=notifications&email_token=AACRDMQYZFZSOBHD25QWAX3RC3SEVA5CNFSM4KGIIFWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL2BNNQ#issuecomment-586421942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACRDMSTBPY7ZBXZQXAJ6QLRC3SEVANCNFSM4KGIIFWA>
> .
>

tomjenkinson commented:
Thanks @starksm64 and @scottmarlow

tomjenkinson commented:
retest this please

tomjenkinson commented:
I have disabled the build of glassfish if and running of TCK unless api/src/main/java/jakarta/transaction/Transaction.java does not exist as the relevant changes to those projects have not been made yet

tomjenkinson commented:
Thanks!

tomjenkinson commented:
Issue to follow up with renabling of the parts of CI temporarily removed: https://github.com/eclipse-ee4j/jta-api/issues/85
